Don't use NewSomething when you refactor (#135)
andycroll/andycroll.com: _posts/2021/2021-08-02-dont-call-replacement-classes-new.md

---

> title: "Don't call a new verison of something new"

Reviewer: *version

description: "A opinionated naming subtlety"
layout: article
category: ruby
image:
  base: '2020/calculate-a-mean-average-from-a-ruby-array'
  alt: ""
  credit: 
  source: "https://unsplash.com/photos/3xwrg7Vv6Ts"

---

> Often in long-lived applications we might want to rewrite or refactor the approach that we have previously taken.

Reviewer: ```suggestion
In long-lived applications, we often want to rewrite or refactor an approach that we have previously taken.
```

## Instead of...

> ...using 

Reviewer: *using:

```ruby
class NewDoThingJob << ActiveJob
  # new implementation
end
```

or

```ruby
class DoThingJobVersionTwo << ActiveJob
  # new implementation
end
```

## Choose...

to rename or namespace the existing code. And create the replacement version with the correct future naming.

```ruby
module Removing
  class DoThingJob << ActiveJob
    # existing implementation
  end
end
```

```ruby
class DoThingJob < Removing::DoThingJob
```

Then change all the places the non-namespaced version is called. Then deploy and ensure that no jobs are still enqueued with the non-namespaced version.

Now you're free to reimplement in the correct (and future) location.

```ruby
class DoThingJob
  # new implmentation
end
```

> Once you've built, tested and changed all the calls to the new version you can delete the (now unused) namespaced version.

Reviewer: The code examples were not the easiest to follow this time around. Not sure _where_ things were going. Had to reread multiple times to be clear on the refactoring process you're describing and I'm still not sure I've fully grasped it.

Reviewer: No Oxford comma? 🥺

Reviewer: *, and updated...

## Why?

> This is a habit that helps communicate intention.

Reviewer: *communicate intent

> These refactoring projects are often large in scale and easy to make mistakes. But this methodology clearly lays out the intention to replace this existing code.

Reviewer: Maybe rather than highlighting making mistakes it's better to highlight the fact that it's not always easy to see what's going on and why...

> If you call your new implmentation `NewSomething` it's not easy for other folks in your team (and future you) to understand the direction of development of the code. "Should I use the New version?"

Reviewer: *implementation

Reviewer: *the new version yet?

> You also won't be left with lots of `NewWhatever` objects littering your code when you inevitably don't do the final renaming.

Reviewer: Consistency with `NewSomething` earlier?

And maybe suggestion that renaming might not end up happening, at least for a while, instead of the "when you inevitably don't do it..." Unless you wanna keep the sass!

> Picking a single namespace for this kind of refactoring also gives you one place to look for all incomplete refactorings and unused code in your application.

Reviewer: Y'see it wasn't also clear that `Removing` was a genuine suggestion for a namespace.

## Why not?

> There's no "code based" reason to do this, as this is a personal preference, plus it adds a little addtional ceremony in exchange for reducing errors.

Reviewer: This sentence doesn't read very well.

Maybe just: This is merely a personal preference. There's no code-based reason to do this.

*additional (but not sure if you want/need that part)

> You also shouldn't blindly copy existing naming. Perhaps there's a better word choice that more accurately reflects what the new code is doing?

Reviewer: Think this point could be made more clearly. Couple reads before I got what you were saying.

> If you're building an API consumed by customers/user outside your team this might not be a good solution. If you're developing a new version, you'll still likely to need to support the existing version. In this case a new `v2` namespace is a good idea.

Reviewer: *users

---

> title: "Don't call a new version of something 'new' when refactoring"

Author: ```suggestion
title: "Don’t call a new version of something ”new” when refactoring"
```

description: "A opinionated naming subtlety"
layout: article
category: ruby
image:
  base: '2021/dont-call-replacement-classes-new'
  alt: "New"
  credit: Nick Fewings
  source: "https://unsplash.com/photos/1SsUquHPNT8"

---

> In long-lived applications, we often want to rewrite or refactor an approach that we have previously taken.

Reviewer: *an approach _to do a particular thing_ (not super keen on that and maybe it's not needed..)

Author: ```suggestion
In long-lived applications, we often want to rewrite or refactor an approach that we have previously taken to solving a problem.
```

> Here's a, non-obvious, naming technique I've used during refactoring to ensure that the improved code ends up in well named classes and any old implementation is clearly marked for deletion.

Reviewer: *well-named

*..and that any old implementations are clearly marked for deletion.

Author: ```suggestion
Here's a, non-obvious, naming technique I've used during refactoring to ensure that the improved code ends up in well-named classes and that any old implementation are clearly marked for deletion.
```

> You might find this helps organise your work or communicate the intent of future changes with your teammates.

Reviewer: *that this helps to organise...or better communicate...

Author: ```suggestion
You might find this helps to organise your work or better communicate the intent of future changes with your teammates.
```

## Instead of...

...littering your codebase with the word "new":

```ruby
class NewDoThingJob << ActiveJob
  # new implementation
end
```

or

```ruby
class DoThingJobVersionTwo << ActiveJob
  # new implementation
end
```

## Choose...

> to rename or, better, namespace the existing code to mark it as deprecated. Then create the replacement version with the correct future naming.

Reviewer: Yesss. Much better!!

```ruby
module ToBeRemoved
  class DoThingJob << ActiveJob
    # existing implementation
  end
end
```

```ruby
class DoThingJob < ToBeRemoved::DoThingJob
```

> Then change all the places the non-namespaced version is called. Then deploy and ensure that no jobs are still enqueued with the non-namespaced version.

Reviewer: This part still isn't so clear. Maybe be more explicit about what is being changed where and to what?

Author: ```suggestion
Where the non-namespace implementation is called, change it to call the implementation in the deprecated namespace. Then deploy and ensure that the application is working as expected and that the non-namespaced version of the code is not called or enqueued anywhere.
```

> Now you're free to reimplement in the correct (and future) location.

Reviewer: Could be worded a bit better.

Author: ```suggestion
Now you can implement the new, improved approach in the correctly named location.
```

```ruby
class DoThingJob
  # new implmentation
end
```

Once you've built, tested and changed all the calls to the new version you can delete the (now unused) namespaced version.

## Why?

This is a habit that helps communicate intent.

> These refactoring projects are often large in scale and it can be hard to see what's going on in the midst of a series of changes. But this methodology clearly lays out the intention to replace this existing code.

Reviewer: Remove the 'but'.

*intention to replace existing code

Author: ```suggestion
These refactoring projects are often large in scale and it can be hard to see what's going on in the midst of a series of changes. This methodology clearly lays out the intention to replace existing code.
```

If you call your new implementation `NewSomething` it's not easy for other folks in your team (and future you) to understand the direction of development of the code. "Should I use the new version yet?"

> You also won't be left with lots of `NewSomething` objects littering your code when you inevitably don't do the final renaming. :-)

Reviewer: *You also don't run the risk of being left with.. (suggestion)

Or

*Furthermore, you won't be left with..

Author: ```suggestion
Furthermore, you don't run the risk of being left with lots of `NewSomething` objects littering your code when you inevitably don't do the final renaming. :-)
```

Picking a single namespace for this kind of refactoring also gives you one place to look for all incomplete refactorings and unused code in your application.

> This staggered renaming also lets you deploy changes in a series of smaller releases and avoids the creation of long running refactoring branches that can be a nightmare to rebase, merge and deploy.

Reviewer: *long-running

No Oxford comma? 🥺

Author: ```suggestion
This staggered renaming also lets you deploy changes in a series of smaller releases and avoids the creation of long-running refactoring branches that can be a nightmare to rebase, merge, and deploy.
```
